databases/week3/mongo.txt

```diff
@@ -0,0 +1,213 @@
+---------------------------------------------------------------- Get meals that has a price smaller than a specific price fx 90
+
+db.Meal.find({price: {$lt: 50}})
+
+----------------------------------------------------------------Get meals that still has available reservations
+
+db.Meal.updateMany({}, {$set: {status: "available"}})
+db.Meal.find({status: { $eq: "available"}})
+
+------------------------------------------------------------------ Get meals that partially match a title. Rød grød med will match the meal with the title Rød grød med fløde
+
+db.Meal.find( { meal_titl: { $in: [ /^Sushi/] } } )
+
+----------------------------------------------------------------------- Get meals that has been created between two dates
+
+db.Meal.updateOne({_id: 1}, {$set: {created_date: new ISODate("2022-04-18")}})
+db.Meal.updateOne({_id: 2}, {$set: {created_date: new ISODate("2022-08-28")}})
+
+
+db.Meal.find({created_date: {$gte:ISODate("2022-03-01"), $lt:ISODate("2022-06-01")}})
+
+------------------------------------------------------------------------- Get only specific number of meals fx return only 5 meals
+
+db.Meal.find().limit(1)
+
+------------------------------------------------------------------------------Get the meals that have good reviews
+
+db.Meal.aggregate( [
+    {
+        $lookup:
+        {
+            from: "Review",
+            localField: "_id",
+            foreignField: "meal_id",
+            as: "stars"
+        }
+    },
+
+    {
+        $project:
+        {
+            "_id": 1,
+            "meal_titl": 1,
+            "location": 1,
+            "review_titl": 1,
+            "stars": "$stars.stars"
+        }
+    },
+
+    {
+    $match: { stars: { $gt: 5 } }
+    },
+
+    {
+        $unwind: "$stars"
+    }
+    
+] )
+
+------------------------------------------------------------------------Get reservations for a specific meal sorted by created_date
+
+db.Reservation.updateOne({_id: 1}, {$set: {created_date: new ISODate("2022-04-18")}})
+
+db.Reservation.insertOne(
+    {
+        "_id": 3,
+        "number_of_guests": 2,
+        "meal_id": 1,
+        "created_date": new ISODate("2023-02-18"),
+        "contact_phonenumber": 333333333,
+        "contact_name": "Jack",
+        "contact_email": "jack@gmail.com"
+    }
+)
+
+db.Reservation.find({meal_id: 1}).sort({created_date: -1})
+
+----------------------------------------------------------------------------Sort all meals by average number of stars in the reviews
+
+db.Meal.aggregate( [
+    {
+        $lookup:
+        {
+            from: "Review",
+            localField: "_id",
+            foreignField: "meal_id",
+            as: "stars"
+        }
+    },
+
+    {
+        $project:
+        {
+            "_id": 1,
+            "meal_titl": 1,
+            "location": 1,
+            "review_titl": 1,
+            "stars": "$stars.stars"
+        }
+    },
+
+    {
+      $sort: { stars: 1 }
+    },
+
+    {
+        $unwind: "$stars"
+    }
+    
```

Comment:
You can modify this query to calculate an average rating:
db.Meal.aggregate( [
  {
    $lookup: {
      from: "Review",
      localField: "_id",
      foreignField: "meal_id",
      as: "reviews"
    }
  },
  {
    $project: {
      "_id": 1,
      "meal_titl": 1,
      "location": 1,
      "review_titl": 1,
      "stars": "$reviews.stars"
    }
  },
  {
    $unwind: "$stars"
  },
  {
  $match: {
    "reviews": { $ne: [] }   //this is ensures that meals without reviews are excluded from the result
  }
},
  {
    $addFields: {
      "avgStars": { $avg: "$stars" }  // calculate average rating but returns null value for meals without rating (if $match hasn't been added)
    }
  },
  {
    $sort: { "avgStars": -1 }
  }
] )